fix(http): rework HttpParams to always be immutable (#20430)

rfuhrer opened the pull request:
I reworked the HttpParams object so that it's a simpler immutable object.  It used to track it's current map and then it's pending updates in separate Maps.  This might make sense if the object wasn't immutable, but since it is, why not just return the updated map every time?  

The object looked immutable from the outside as the modification operations (append, set, delete) would return a new object, however as soon as you called a read operation (get, keys, toString) it would actually modify the object in order to spit out the correct values. Due to a bug in this logic it would create duplicate parameters in certain cases (the issue referenced)

The rework simplifies everything by removing the tracked `clonefrom` and `updates` and just keeping the existing `map` updated right when you do an operation that results in a clone.  This way read operations are already working with a fully up to date map and can just return the values requested

PR Close #20430 

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[x] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
Currently when you make a read operation on a HttpParam object, it modifies the object.  This is contrary to the documentation that says the object is immutable.

Issue Number: #20430 


## What is the new behavior?
All change operations now correctly return a new completed object, and all reads return values from this object without modifying it

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

Commit:
fix(http): rework HttpParams to always be immutable (#20430)

rfuhrer commented:
Signed the CLA

aciccarello commented:
Wouldn't it be a lot simpler to simply clean up `this.updates` on init rather than reworking the entire class?

rfuhrer commented:
Simpler to just get a working fix, but it leaves the class needlessly complicated and inefficient.

It currently saves an entire copy of the object where it was cloned from just so it can know what kind of "state" it's in when running it's update.

It saves a "map" that contains the parameters since you last called a read operation, and then a "updates" object that contains everything new that has happened but hasn't been read yet. Then it needs to track all of these updates (eg. Removing an added parameter if you call remove before doing a read) which is just as much work as just keeping the actual map up to date anyways

Finally the class is supposed to be immutable and yet doing any read actually modifies the object. That's not really immutable.

My rework is almost exclusively deleting code, I didn't really rewrite anything. I think the end result is worth the little bit of extra testing and review vs just riding the init bug and calling it a day. Smaller, simpler, less memory usage, and easier to maintain >> smaller pull request imho :)

jmdobry commented:
I'm currently afflicted by this bug. I started refactoring away from `URLSearchParams` (which is deprecated),  only to find that `HttpParams` isn't actually immutable, as every time I do a read my parameters get duplicated on the params, so now I'm reverting back to `URLSearchParams`...

rfuhrer commented:
@jmdobry I feel you.  I'm just lucky my server basically ignores the duplicates so it doesn't effect me too much.  But it seems like a pretty major bug that really effects certain people.  I was hoping this would have been looked at months ago, but I didn't want to be the annoying guy prodding the angular team every 3 days wondering why they haven't looked at this yet.

alfaproject commented on packages/common/http/src/params.ts:
```diff
@@ -113,45 +111,35 @@ export class HttpParams {
       });
     } else {
-      this.map = null;
+      this.map = new Map<string, string[]>();
```
```ts
// ...
} else {
  this.map = new Map<string, string[]>();
  if (options.fromObject) {
    // ...
  }
}
``

rfuhrer replied:
Good catch 👍

rfuhrer replied:
Good catch 👍

alfaproject replied:
You still didn't optimise this d:

rfuhrer replied:
I definitely read this comment wrong and thought you fixed a part where I missed setting map to not be null (hence the "good catch" comment).

But we left a bug in Angular for 4 more months because of a single line of code _could_ be removed but doesn't affect anything? I would also argue for it to be left the same way because it's more apparent that there's 3 cases that can happen when constructing a HttpParams object.

alexxxnf commented:
Is there any chance for this PR to get merged anytime soon?

rfuhrer commented:
![image](https://user-images.githubusercontent.com/2529744/41481803-922cb7b2-7090-11e8-9524-2c7f8174de09.png)
😞

rfuhrer commented:
OK, I added the change @alfaproject requested and rebased the project on master so the build server would actually work (who knew that 8 months of changes to the build environment would break your PR even though the code was fine 😄).  At this point I don't think there's anything more I can do to get this thing accepted.  Other than rebasing again in 8 months...

rfuhrer commented:
Since I was digging back into this code it made me realize something funny that's also wrong with the old implementation.

https://stackblitz.com/edit/angular-arkubk?file=src%2Fapp%2Fapp.component.ts

By keeping this "updates" map, your object constantly gets bigger and bigger the longer you keep it around and set/append/delete things on it. So if you're like me and have a search screen with filters and you're storing these filters in the HttpParams object since that's what gets sent to the server anyways, the more filters a user clicks/unclicks the more memory you use.

mhevery commented:
Was this addressed in https://github.com/angular/angular/pull/29045?

JoostK commented:
@mhevery Yes, #29045 has landed as fix for the original issue without moving away from the lazy datastructures as implemented in this change.

rfuhrer commented:
Yeah the original bug  #20430 was already fixed.  I'll close this
